build: add scaffolding to generate `math/special/*` packages

gunjjoshi opened the pull request:
Resolves None.

## Description

> What is the purpose of this pull request?

This pull request:

-   adds scaffold templates and scripts for generating packages in `math/special/*`.

## Related Issues

> Does this pull request have any related issues?

None.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

gunjjoshi committed:
chore: initial scaffold setup

gunjjoshi committed:
chore: initial scripts

gunjjoshi committed:
refactor: update scripts

gunjjoshi committed:
refactor: update script

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi committed:
refactor: update scripts and templates

gunjjoshi committed:
refactor: remove unwanted files

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/dtype_ndarray_mappings.json:
We have this new file now, that contains the mappings which will be used in `index.d.ts`, such as here: https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/special/abs/docs/types/index.d.ts

gunjjoshi committed:
chore: generate acosd

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: skipped
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: missing_dependencies
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

gunjjoshi committed:
chore: generate acosd

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: skipped
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/runner.js:
```diff
@@ -0,0 +1,462 @@
+	// for ( i = 0; i < dataKeys.length; i++ ) {
+
+		// dataKey = dataKeys[ i ];
+		dataKey = '@stdlib/math/base/special/acosd';
```
For testing initially, I didn't generate all of the packages using the `for` loop. Instead, I have hardcoded `acosd`, to generate it.

gunjjoshi committed:
refactor: use correct scalar function name in C

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: skipped
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi commented:
Getting some CI failures which might be arising from [`@stdlib/ndarray/base/unary`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/ndarray/base/unary), specifically from here: https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/ndarray/base/unary/src/c_d_as_z_d.c . Maybe, it must have `stdlib_complex128_t` instead of `stdlib_complex64_t`?

kgryte commented:
@gunjjoshi Yes, I'll try and take a look at `ndarray/base/unary` tonight.

gunjjoshi committed:
refactor: add handlebars for output policy phrases

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into scaffolds

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/data/README__md.txt:
```diff
@@ -0,0 +1,229 @@
+
+The function accepts the following options:
+
+-   **dtype**: output ndarray [data type][@stdlib/ndarray/dtypes]. Must be a {{OUTPUT_POLICY_PHRASE}} [data type][@stdlib/ndarray/dtypes].
```
```suggestion
-   **dtype**: output ndarray [data type][@stdlib/ndarray/dtypes]. Must be {{OUTPUT_POLICY_PHRASE}} [data type][@stdlib/ndarray/dtypes].
```

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/data/README__md.txt:
```diff
@@ -0,0 +1,229 @@
+});
+// returns <ndarray>
+
+var dt = getDType( y );
```
```suggestion
var dt = String( getDType( y ) );
```

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/data/README__md.txt:
```diff
@@ -0,0 +1,229 @@
+
+## Notes
+
+-   The output data type [policy][@stdlib/ndarray/output-dtype-policies] only applies to the main function and specifies that, by default, the function must return an [ndarray][@stdlib/ndarray/ctor] having a {{OUTPUT_POLICY_PHRASE}} [data type][@stdlib/ndarray/dtypes]. For the `assign` method, the output [ndarray][@stdlib/ndarray/ctor] is allowed to have any supported output [data type][@stdlib/n ...
```
```suggestion
-   The output data type [policy][@stdlib/ndarray/output-dtype-policies] only applies to the main function and specifies that, by default, the function must return an [ndarray][@stdlib/ndarray/ctor] having {{OUTPUT_POLICY_PHRASE}} [data type][@stdlib/ndarray/dtypes]. For the `assign` method, the output [ndarray][@stdlib/ndarray/ctor] is allowed to have any supported output [data type][@stdlib/ndarray/dtypes].
```

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/data/README__md.txt:
```diff
@@ -0,0 +1,229 @@
+limitations under the License.
+
+-->
+
```
```suggestion

<!-- THIS IS A GENERATED FILE. DO NOT EDIT DIRECTLY. -->

```

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/data/docs/repl__txt.txt:
```diff
@@ -0,0 +1,67 @@
+    Parameters
+    ----------
+    x: ndarray
+        Input array. Must have a numeric or "generic" data type.
```
@gunjjoshi Looks like we need an "input policy phrase" here. "numeric" is not true for all apis (e.g., those not supporting complex input dtypes).

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/data/docs/repl__txt.txt:
```diff
@@ -0,0 +1,67 @@
+        Options.
+
+    options.dtype: string|DataType (optional)
+        Output array data type. Must be a real-valued or "generic" data type.
```
```suggestion
        Output array data type. Must be {{OUTPUT_POLICY_PHRASE}} data type.
```

This is a little bit tricky, as this assumes that the line will not exceed 80 chars.

gunjjoshi replied:
I've set this whole line as a handlebar, rather than just the `{{OUTPUT_POLICY_PHRASE}}`, as it was earlier.

kgryte replied:
Yes, that makes sense, as you need to handle line wrapping.

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/data/docs/repl__txt.txt:
```diff
@@ -0,0 +1,67 @@
+    Parameters
+    ----------
+    x: ndarray
+        Input array. Must have a numeric or "generic" data type.
```
Same thing. Input policy phrase.

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/data/docs/types/index__d__ts.txt:
```diff
@@ -0,0 +1,146 @@
+
+/// <reference types="@stdlib/types"/>
+
+import { typedndarray, {{NDARRAY_IMPORT_TYPES}}, genericndarray, RealAndGenericDataType as DataType, Order } from '@stdlib/types/ndarray';
```
This isn't correct. `RealAndGenericDataType` will not be correct for all APIs. `opts.dtype` is subject to the output dtype policy.

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/data/docs/types/index__d__ts.txt:
```diff
@@ -0,0 +1,146 @@
+	* var arr = ndarray2array( y );
+	* // returns {{JS_EXAMPLE_RESULTS_2X2}}
+	*/
+	( x: InputArray, options?: Options ): typedndarray<number>; // FIXME: we lose type specificity here, as the output ndarray data type is determined according to the output data type policy in conjunction with the `dtype` option
```
This isn't correct. E.g., what happens if a unary function returns a complex number? Meaning, `number` is not always correct. I don't have a great solution for this. We may need to do conditional overloads. Meaning, for those apis accepting and returning complex numbers, we may need an overload for when `x` is a complex array, as then the output should also be a complex array. In short, the structure of this file depends on the output dtype policy.

gunjjoshi replied:
I've tried to solve this in https://github.com/stdlib-js/stdlib/pull/8139/commits/121f41f7a4c2bd06ebd09018fe3253bedadbaabb. We have a [json file ](https://github.com/stdlib-js/stdlib/blob/121f41f7a4c2bd06ebd09018fe3253bedadbaabb/lib/node_modules/%40stdlib/_tools/scaffold/math-special-unary/scripts/element_types.json) which maps input/output dtype policies to `number`/`ComplexLike`/`number|ComplexLike`/`boolean`, which is then placed in `genericndarray<...>`/`typedndarray<...>` [here](https://github.com/stdlib-js/stdlib/pull/8139/commits/121f41f7a4c2bd06ebd09018fe3253bedadbaabb#diff-32f6da9f3e67b0ec6d9d60a7fd4c91e039ff6718a7066ad7716ceb58595b121a).

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/data/docs/types/test__ts.txt:
```diff
@@ -0,0 +1,154 @@
+		'dtype': 'float64'
+	});
+
+	{{ALIAS}}( x ); // $ExpectType typedndarray<number>
```
Same thing. The output array may or may not contain numbers depending on the API and the input array dtype.

gunjjoshi replied:
Applied a similar thing as [above](https://github.com/stdlib-js/stdlib/pull/8139#discussion_r2572847967) via https://github.com/stdlib-js/stdlib/pull/8139/commits/81316cad2b248c10f3a899f4ba95114b63f1241c.

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/data/docs/types/test__ts.txt:
```diff
@@ -0,0 +1,154 @@
+		'dtype': 'float64'
+	});
+
+	{{ALIAS}}.assign( x, x ); // $ExpectType float64ndarray
```
This works so long as an api accepts floating-point ndarrays. That may not always be true. The use of `float64` throughout this file and examples assumes floating-point support.

gunjjoshi replied:
Updated this [here](https://github.com/stdlib-js/stdlib/pull/8139/commits/7330ac4ccfffd935078ba308a7877daee60a16a1) to use the [`primary_dtype`](https://github.com/stdlib-js/stdlib/pull/8538) field that we added in the function database.

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/data/README__md.txt:
```diff
@@ -0,0 +1,229 @@
+var ndarray2array = require( '@stdlib/ndarray/to-array' );
+var array = require( '@stdlib/ndarray/array' );
+
+var x = array( {{JS_EXAMPLE_VALUES_2X2}} );
```
This is a little tricky, as it assumes that `x` is allowed to be a `float64` array. I am not sure that will always be the case. E.g., we could have a ufunc which only accept integer arrays or complex number arrays, etc. I am not sure the most robust solution here.

kgryte replied:
Can you confirm that, for all ufuncs in math special, we will support float64, float32, and generic?

gunjjoshi replied:
Yes, I believe all of them will support `float64`, `float32` and generic. And for `cpolar`, which we discussed in the call, I found that we can implement `polar`, by keeping the phase angle as 0 or `pi`. MATLAB already supports both complex and real numbers for its `angle()` function. The documentation [here](https://in.mathworks.com/help/matlab/ref/angle.html) states:

"When the elements of z are non-negative real numbers, angle returns 0. When the elements of z are negative real numbers, angle returns π."

kgryte replied:
Yes, but then we'd need real-valued scalar kernels to explicitly handle this.

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/data/test/test__main__js.txt:
```diff
@@ -0,0 +1,284 @@
+	expected = map( x, naryFunction( base, 1 ) );
+	t.deepEqual( ndarray2array( y ), ndarray2array( expected ), 'returns expected value' );
+
+	t.strictEqual( getDType( y ), 'generic', 'returns expected value' );
```
```suggestion
	t.strictEqual( String( getDType( y ) ), 'generic', 'returns expected value' );
```

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/data/test/test__main__native__js.txt:
```diff
@@ -0,0 +1,293 @@
+	expected = map( x, naryFunction( base, 1 ) );
+	t.deepEqual( ndarray2array( y ), ndarray2array( expected ), 'returns expected value' );
+
+	t.strictEqual( getDType( y ), 'generic', 'returns expected value' );
```
```suggestion
	t.strictEqual( String( getDType( y ) ), 'generic', 'returns expected value' );
```

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/dtype_ndarray_mappings.json:
```diff
@@ -0,0 +1,12 @@
+{
+	"input_dtypes": {
```
As discussed during a previous call, you can build out this table for completeness to include other dtype mappings.

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/generate_files.js:
```diff
@@ -0,0 +1,455 @@
+	}
+
+	result = JSON.stringify( typeEnums ) + '\n';
+	return result;
```
```suggestion
	var i;

	typeEnums = [];
	for ( i = 0; i < matches.length; i++ ) {
		typeEnums.push( str2enum( matches[ i ][ 0 ] ) );
		typeEnums.push( str2enum( matches[ i ][ 1 ] ) );
	}

	return JSON.stringify( typeEnums ) + '\n';
```

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/generate_files.js:
```diff
@@ -0,0 +1,455 @@
+			matchIndex = 0;
+			for ( k = 0; k < matches.length; k++ ) {
+				if ( matches[ k ][ 0 ] === inputType &&
+					matches[ k ][ 1 ] === outputTypes[ j ] ) {
```
```suggestion
				if ( matches[ k ][ 0 ] === inputType && matches[ k ][ 1 ] === outputTypes[ j ] ) { // eslint-disable-line max-len
```

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/generate_files.js:
```diff
@@ -0,0 +1,455 @@
+	matches = result.reorderedMatches;
+
+	cOut = header;
+	cOut += Object.keys( uniqueIncludes ).join( '\n' ) + '\n';
```
@gunjjoshi Use `@stdlib/utils/keys`.

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/generate_files.js:
```diff
@@ -0,0 +1,455 @@
+
+		cOut += '\t// ' + inputType + ' (' + outputTypes.length + ')\n';
+		for ( j = 0; j < outputTypes.length; j++ ) {
+			cOut += '\tSTDLIB_NDARRAY_' + matches[ functionIndex ][ 0 ].toUpperCase() + ', STDLIB_NDARRAY_' + matches[ functionIndex ][ 1 ].toUpperCase();
```
@gunjjoshi Use uppercase string utility. `stdlib/string/base/uppercase`.

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/generate_files.js:
```diff
@@ -0,0 +1,455 @@
+
+	// Generate types.js:
+	jsOut = generateTypesFile( matches, header );
+	writeFileSync( join( destDir, 'lib/types.js' ), jsOut, {
```
You make heavy use of string concatenation to generate entire files, including license headers. I wonder if we could use template files with handlebar template strings `{{}}` similar to what we do for other files. That might reduce the visual noise in this file.

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/output_policy_phrases.json:
```diff
@@ -0,0 +1,7 @@
+  "floating_point_and_generic": "floating-point or \"generic\"",
+  "numeric_and_generic": "numeric or \"generic\"",
+  "real_and_generic": "real-valued or \"generic\"",
+  "real_floating_point_and_generic": "real-valued floating-point or \"generic\""
```
```suggestion
  "floating_point_and_generic": "a floating-point or \"generic\"",
  "numeric_and_generic": "a numeric or \"generic\"",
  "real_and_generic": "a real-valued or \"generic\"",
  "real_floating_point_and_generic": "a real-valued floating-point or \"generic\""
```

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/runner.js:
```diff
@@ -0,0 +1,479 @@
+// FUNCTIONS //
+
+/**
+* Returns a string containing environment variable for generating pseudorandom numbers.
```
```suggestion
* Returns a string containing an environment variable for generating pseudorandom numbers.
```

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/runner.js:
```diff
@@ -0,0 +1,479 @@
+		v = values[ i ];
+		results = func( v );
+		results = round( results * 100 ) / 100; // Rounding results to 2 decimal places
+		vals.push( '~' + results.toString() );
```
I think you only want to round if the result is not exact (i.e., an integer value). Otherwise, if results are exact, you don't need a `~`. Furthermore, if integer-valued, you also need to include trailing `.0` characters to the results if the array type is floating-point.

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/runner.js:
```diff
@@ -0,0 +1,479 @@
+
+/**
+* Returns the input ndarray type environment variable.
+* @private
```
```suggestion
*
* @private
```

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/runner.js:
```diff
@@ -0,0 +1,479 @@
+
+	// Using the scalar kernels to compute the dependencies:
+	scalarKernels = funcData.scalar_kernels;
+	keys = Object.keys( scalarKernels );
```
Use `utils/keys`.

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/runner.js:
```diff
@@ -0,0 +1,479 @@
+
+	for ( i = 0; i < keys.length; i++ ) {
+		kernelPath = scalarKernels[ keys[ i ] ];
+		if ( kernelPath.startsWith( '/' ) ) {
```
Use `string/[base/]starts-with`.

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/runner.js:
```diff
@@ -0,0 +1,479 @@
+*
+* @example
+* var envVar = outputPolicyPhraseEnvVar( 'floating_point_and_generic' );
+* // returns 'floating-point or \"generic\"'
```
```suggestion
* // returns 'a floating-point or \"generic\"'
```

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/runner.js:
```diff
@@ -0,0 +1,479 @@
+		pkgDesc = 'Compute the ' + o.short_desc + ' for each element in an ndarray.';
+		moduleDesc = 'Compute the ' + o.short_desc + '.';
+		mainDesc = 'Computes the ' + o.short_desc + ' for each element in an ndarray.';
+		testDesc = 'computes the ' + o.short_desc + ' for each element in an ndarray';
```
Remind me: what is our plan for functions like `round`?

gunjjoshi replied:
I think we discussed to handle such cases by having a dedicated field for description in our [function database](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/special/data/unary_function_database.json). If, for any package, its database entry has a field for "description", then these descriptions would have to be overridden and the one mentioned in the function database will be used.

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/runner.js:
```diff
@@ -0,0 +1,479 @@
+			keywords = keywords.concat( o.extra_keywords );
+		}
+		envs.push( 'KEYWORDS=\'' + keywords.join( ',' ) + '\'' );
+		envs.push( 'YEAR=\'' + new Date().getFullYear() + '\'' );
```
Use `time/current-year`.

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/script.js:
```diff
@@ -0,0 +1,313 @@
+			/*
+			* For the first iteration:
+			*
+			* allowedCasts = [
```
'float32' and 'float64' are not mostly safe casts for 'complex64'

gunjjoshi replied:
Removed them and updated related comments below it, thanks for catching!

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/script.js:
```diff
@@ -0,0 +1,313 @@
+		// Remove the dtypes which are not allowed for output, according to the output policy:
+		filtered = [];
+		for ( k = 0; k < allowedCasts.length; k++ ) {
+			if ( odt.indexOf( allowedCasts[ k ] ) !== -1 ) {
```
Prefer using `array/base/assert/contains` rather than `indexOf` here and elsewhere when searching for needles in arrays.

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/script.js:
```diff
@@ -0,0 +1,313 @@
+
+			// Extract the function name from the scalar kernel path and create C function name:
+			cFunctionName = scalarKernel.split( '/' ).pop(); // Get last part (Ex: 'acosd')
+			cFunctionName = 'stdlib_base_' + cFunctionName;
```
This isn't always correct. E.g., `stdlib_base_float64_identity`.

gunjjoshi replied:
@kgryte Will we have scalar kernels from any other namespace apart from `@stdlib/number/*` and `@stdlib/math/base/special/*`? If no, then these 2 can be handled easily. But if yes, then we might need to think of some other way to form the C scalar kernel name. This is because both of these namespaces follow different naming patterns, for example:

- `@stdlib/math/base/special/abs` -> `stdlib_base_abs` (the word after "base", i.e., "special" is skipped)
- `@stdlib/number/uint8/base/identity` -> `stdlib_base_uint8_identity` (here we don't have any word after "base" to skip, as we need to keep "identity" in the signature, along with the word before "base", i.e., "uint8")

One possible idea I can think of is to pick up the C scalar kernel name from the `README.md` of that particular package, such as from here: https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/number/uint16/base/identity/README.md?plain=1#L130

kgryte replied:
Also `@stdlib/complex/*`.

kgryte replied:
As discussed on call, let's go ahead and explore adding a `__stdlib__.scaffold.alias_prefix` meta data field (e.g., `stdlib_base_complex64_`) to parent namespaces.

gunjjoshi replied:
Updated the script to use `__stdlib__.scaffold.alias_prefix` via https://github.com/stdlib-js/stdlib/pull/8139/commits/9f3d03f763f680635ec1f0b2e67938e0cda006a7

kgryte commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/runner.js:
```diff
@@ -0,0 +1,479 @@
+		envs.push( 'PKG_PATH=\'stdlib/math/special\'' );
+
+		pkgDesc = 'Compute the ' + o.short_desc + ' for each element in an ndarray.';
+		moduleDesc = 'Compute the ' + o.short_desc + '.';
```
```suggestion
		moduleDesc = 'Compute the ' + o.short_desc + ' for each element in an ndarray.';
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (changes requested):
Left an initial review. Main concerns revolve around edge cases.

gunjjoshi committed:
Merge branch 'stdlib-js:develop' into scaffolds

gunjjoshi committed:
refactor: use input policy phrase in repl

gunjjoshi committed:
refactor: use stdlib packages

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi committed:
refactor: use stdlib packages

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi committed:
refactor: use array/base/assert/contains

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi committed:
chore: remove scripts field

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi committed:
fix: lint error, check exit code directly

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: missing_dependencies
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi committed:
refactor: handle example results correctly

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi committed:
feat: add more dtype mappings

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi committed:
docs: remove invalid dtypes from comments

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi committed:
feat: add support for OUTPUT_DTYPE_TS

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: passed
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi committed:
Merge branch 'stdlib-js:develop' into scaffolds

gunjjoshi committed:
feat: add support to handle prefix aliases of different namespaces

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

gunjjoshi committed:
Merge branch 'stdlib-js:develop' into scaffolds

gunjjoshi committed:
refactor: update dtype-to-ndarray type mappings

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi committed:
feat: add support for dynamic typescript types

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: passed
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi committed:
feat: add support for dynamic typescript types

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi committed:
feat: add support for primary dtype in test.ts

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: passed
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

gunjjoshi committed:
Merge branch 'stdlib-js:develop' into scaffolds

gunjjoshi committed:
refactor: modularize scripts

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: passed
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi committed:
Merge branch 'stdlib-js:develop' into scaffolds

gunjjoshi committed:
test: use is-almost-equal utilities

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi committed:
feat: add complex value support to example generation functions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into scaffolds

gunjjoshi committed:
feat: add separate README template for complex values

gunjjoshi committed:
feat: add support for complex examples as a separate template

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: passed
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi committed:
Merge branch 'stdlib-js:develop' into scaffolds

gunjjoshi commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

gunjjoshi committed:
feat: add support for skipping packages without C implementations

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/lib/runner.js:
```diff
@@ -0,0 +1,98 @@
+	dataKey = '@stdlib/math/base/special/acosd';
+	o = DATA[ dataKey ];
+
+	if ( !hasCImplementation( o.base_alias ) ) {
+		log( 'Skipping @%s: Missing C implementations for one or more kernels.', 'stdlib/math/special/' + o.alias );
```
Now this script skips the creation of those ufuncs which do not have C implementation for all of their scalar kernels.

gunjjoshi commented on lib/node_modules/@stdlib/_tools/scaffold/math-special-unary/scripts/lib/has_c_implementation.js:
```diff
@@ -0,0 +1,135 @@
+		return false;
+	}
+
+	kernelPaths = unique( objectValues( funcData.scalar_kernels ) );
```
Taking only unique scalar kernels here. This helps us to avoid redundant searches, in cases where multiple dtypes use the same kernel. For example, if, in case of `sin`, if`float64` and `generic` use the same `sin` kernel, we should not look up the package.json of `sin` twice.

kgryte committed:
Merge branch 'develop' into scaffolds

kgryte committed:
Merge branch 'develop' into scaffolds
